File: ticket-booking-system/src/main/java/com/ticket/ticket_booking_system/service/SeatService.java
package com.ticket.ticket_booking_system.service;

import java.util.List;
import java.util.UUID;

import com.ticket.ticket_booking_system.entity.Seat;

public interface SeatService {
    
    /**
     * Generate seats for an event by duplicating venue seats
     */
    void generateSeatsForEvent(UUID venueId, UUID eventId);
    
    /**
     * Get all seats for an event
     */
    List<Seat> getSeatsByEvent(UUID eventId);
    
    /**
     * Get available seats for an event
     */
    List<Seat> getAvailableSeatsByEvent(UUID eventId);
    
    /**
     * Toggle seat blocking status
     */
    void toggleBlock(UUID seatId, boolean block);
    
    /**
     * Hold seats for a user temporarily
     */
    void holdSeats(List<UUID> seatIds, UUID userId, int holdDurationMinutes);
    
    /**
     * Reserve seats (mark as unavailable)
     */
    void reserveSeats(List<UUID> seatIds);
    
    /**
     * Release seat holds for a user
     */
    void releaseSeatHolds(UUID userId);
    
    /**
     * Release all expired holds automatically
     */
    void releaseExpiredHolds();
    
    /**
     * Get seat by ID
     */
    Seat getSeatById(UUID seatId);
    
    /**
     * Get seat availability statistics for an event
     */
    SeatAvailabilityStats getAvailabilityStats(UUID eventId);
    
    public static class SeatAvailabilityStats {
        private long totalSeats;
        private long availableSeats;
        private long bookedSeats;
        private long heldSeats;
        private long blockedSeats;
        
        public SeatAvailabilityStats(long totalSeats, long availableSeats, long bookedSeats, long heldSeats, long blockedSeats) {
            this.totalSeats = totalSeats;
            this.availableSeats = availableSeats;
            this.bookedSeats = bookedSeats;
            this.heldSeats = heldSeats;
            this.blockedSeats = blockedSeats;
        }
        
        // Getters and setters
        public long getTotalSeats() { return totalSeats; }
        public void setTotalSeats(long totalSeats) { this.totalSeats = totalSeats; }
        
        public long getAvailableSeats() { return availableSeats; }
        public void setAvailableSeats(long availableSeats) { this.availableSeats = availableSeats; }
        
        public long getBookedSeats() { return bookedSeats; }
        public void setBookedSeats(long bookedSeats) { this.bookedSeats = bookedSeats; }
        
        public long getHeldSeats() { return heldSeats; }
        public void setHeldSeats(long heldSeats) { this.heldSeats = heldSeats; }
        
        public long getBlockedSeats() { return blockedSeats; }
        public void setBlockedSeats(long blockedSeats) { this.blockedSeats = blockedSeats; }
    }
}

File: ticket-booking-admin/src/services/seat.service.ts
import axiosInstance from './api';

export interface Seat {
  seatId: string;
  venueId: string;
  venueName: string;
  eventId: string;
  eventName: string;
  section: string;
  rowNumber: string;
  seatNumber: string;
  seatType: string;
  price: number;
  isAvailable: boolean;
  isBlocked: boolean;
  holdExpiresAt?: string;
  heldByUser?: string;
  createdAt?: string;
}

export interface SeatAvailabilityStats {
  totalSeats: number;
  availableSeats: number;
  bookedSeats: number;
  heldSeats: number;
  blockedSeats: number;
  timestamp: number;
}

export interface SeatUpdateMessage {
  seatId: string;
  section: string;
  rowNumber: string;
  seatNumber: string;
  isAvailable: boolean;
  isBlocked: boolean;
  isHeld: boolean;
  action: 'BLOCKED' | 'UNBLOCKED' | 'HELD' | 'RESERVED' | 'RELEASED';
  timestamp: number;
}

export interface SeatHoldNotification {
  eventId: string;
  message: string;
  type: 'HOLD_CONFIRMED' | 'HOLD_EXPIRED' | 'HOLD_RELEASED';
  expiresAt: number;
}

export interface VenueSeatingConfig {
  layoutType: 'theater' | 'stadium' | 'standard' | 'custom';
  sections: {
    name: string;
    rows: {
      name: string;
      seats: number;
      priceMultiplier: number;
    }[];
  }[];
  basePrice: number;
}

export interface HoldSeatsRequest {
  seatIds: string[];
  holdDurationMinutes?: number;
}

// API Service for Seat Management
export const SeatService = {
  // Get all seats for an event
  getSeatsByEvent: async (eventId: string): Promise<Seat[]> => {
    const response = await axiosInstance.get<Seat[]>(`/api/seats?eventId=${eventId}`);
    return response.data;
  },

  // Get available seats for an event
  getAvailableSeatsByEvent: async (eventId: string): Promise<Seat[]> => {
    const response = await axiosInstance.get<Seat[]>(`/api/seats/available?eventId=${eventId}`);
    return response.data;
  },

  // Get seat availability statistics
  getAvailabilityStats: async (eventId: string): Promise<SeatAvailabilityStats> => {
    const response = await axiosInstance.get<SeatAvailabilityStats>(`/api/seats/stats?eventId=${eventId}`);
    return response.data;
  },

  // Get individual seat details
  getSeatById: async (seatId: string): Promise<Seat> => {
    const response = await axiosInstance.get<Seat>(`/api/seats/${seatId}`);
    return response.data;
  },

  // Admin: Block/unblock a seat
  toggleSeatBlock: async (seatId: string, block: boolean): Promise<void> => {
    await axiosInstance.put(`/api/seats/${seatId}/block?block=${block}`);
  },

  // User: Hold seats temporarily
  holdSeats: async (request: HoldSeatsRequest): Promise<void> => {
    await axiosInstance.post('/api/seats/hold', request);
  },

  // User: Reserve held seats
  reserveSeats: async (seatIds: string[]): Promise<void> => {
    await axiosInstance.post('/api/seats/reserve', seatIds);
  },

  // User: Release user's seat holds
  releaseSeatHolds: async (userId: string): Promise<void> => {
    await axiosInstance.post(`/api/seats/release-holds?userId=${userId}`);
  },

  // Admin: Generate seats for an event from venue layout
  generateSeatsForEvent: async (venueId: string, eventId: string): Promise<void> => {
    await axiosInstance.post(`/api/seats/generate?venueId=${venueId}&eventId=${eventId}`);
  },
};

export default SeatService;